wlroots/xwayland.py

```diff
@@ -0,0 +1,462 @@
+
+    def icccm_input_model(self) -> int:
+        return lib.wlr_xwayland_icccm_input_model(self._ptr)
+
+    @property
+    def surface(self) -> WlrSurface:
+        return WlrSurface(self._ptr.surface)
+
+    @property
+    def x(self) -> int:
+        return self._ptr.x
+
+    @property
+    def y(self) -> int:
+        return self._ptr.y
+
+    @property
+    def width(self) -> int:
+        return self._ptr.width
+
+    @property
+    def height(self) -> int:
+        return self._ptr.height
+
+    @property
+    def override_redirect(self) -> bool:
+        return self._ptr.override_redirect
+
+    @property
+    def mapped(self) -> bool:
+        return self._ptr.mapped
+
+    @property
+    def title(self) -> str | None:
+        return str_or_none(self._ptr.title)
+
+    @property
+    def wm_class(self) -> str | None:
+        # `self._ptr.class` is invalid syntax but this works.
+        return str_or_none(getattr(self._ptr, "class"))
+
+    @property
+    def wm_instance(self) -> str | None:
+        return str_or_none(self._ptr.wm_instance)
+
+    @property
+    def role(self) -> str | None:
+        return str_or_none(self._ptr.role)
+
+    @property
+    def startup_id(self) -> str | None:
+        return str_or_none(self._ptr.startup_id)
+
+    @property
+    def pid(self) -> int:
+        return self._ptr.pid
+
+    @property
+    def parent(self) -> Surface | None:
+        ptr = self._ptr.parent
+        if ptr == ffi.NULL:
+            return None
+        return Surface(ptr)
+
+    @property
+    def window_type(self) -> list[int]:
+        """This is an array of xcb_atom_t."""
+        if self._ptr.window_type_len == 0:
+            return []
+        return ffi.unpack(self._ptr.window_type, self._ptr.window_type_len)
+
+    @property
+    def protocols(self) -> list[int]:
+        """This is an array of xcb_atom_t."""
+        return ffi.unpack(self._ptr.protocols, self._ptr.protocols_len)
+
+    @property
+    def hints_urgency(self) -> int:
+        return self._ptr.hints_urgency
+
+    @property
+    def size_hints(self) -> SizeHints | None:
+        ptr = self._ptr.size_hints
+        if ptr == ffi.NULL:
+            return None
+        return SizeHints(ptr)
+
+    @property
+    def modal(self) -> bool:
+        return self._ptr.modal
+
+    @property
+    def fullscreen(self) -> bool:
+        return self._ptr.fullscreen
+
+    @property
+    def maximized_vert(self) -> bool:
+        return self._ptr.maximized_vert
+
+    @property
+    def maximized_horz(self) -> bool:
+        return self._ptr.maximized_horz
+
+    @property
+    def minimized(self) -> bool:
+        return self._ptr.minimized
+
+    @property
+    def has_alpha(self) -> bool:
+        return self._ptr.has_alpha
+
+    def surface_at(
+        self, surface_x: float, surface_y: float
+    ) -> tuple[WlrSurface | None, float, float]:
+        """A convenience method to match the method of WlrSurface and XdgSurface."""
```

Comment:
Looking at this signature, I feel like `tuple[WlrSurface, float, float] | None` makes more sense, but that can be a coordinated change for all of the `surface_at` functions.